L80: Add a proposal for ClientAsyncResponseReaderInterface destruction.

jacobsa opened the pull request:
See previous discussion in https://github.com/grpc/grpc/pull/26233.

@ctiller I've set you as the approver after the discussion in the thread above. Feel free to re-assign if that's not appropriate. Thanks!

soheilhy commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,123 @@
+which knows it is always allocated on an arena and doesn't care about its
+destructor running. This proposal provides a plan for fixing this so that other
+subclasses can count on their destructor running and their backing memory being
+freed, as with any other C++ class.
```
>This proposal provides a plan for fixing this so that other
> subclasses can count on their destructor running and their backing memory being
> freed, as with any other C++ class.

I would replace this with: 

```
This proposal provides a plan for guaranteed destructor invocation, such that 
subclasses can count on their destructor running and their backing memory being
freed, as with any other C++ class.
```

jacobsa replied:
Done.

soheilhy commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,123 @@
+
+This was introduced by
+[`dd36b153`](https://github.com/grpc/grpc/commit/dd36b15315cd691e86a94d4574bd9f3e3a33633f),
+which introduced the specialization for `grpc::ClientAsyncResponseReader`
```
nit: I would replace `introduced` with `included` to avoid repetition.

jacobsa replied:
Done (except I used `added`).

soheilhy commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,123 @@
+
+*   `ClientAsyncResponseReader` is not the only subclass of the interface. The
+    interface's destructor is virtual, so other subclasses may have destructors
+    with side effects. However they are never actually run! This is a huge
```
> The interface's destructor is virtual, so other subclasses may have destructors
>  with side effects. However they are never actually run! ...

I would replace this with:

```
Even though subclasses can override the virtual destructor to add a side effect, their dtor does not actually run. This results in subtle bugs and confusing behavior, specially because   `delete p.release()` works differently than
 just letting a `std::unique_ptr p` go out of scope.
```

jacobsa replied:
Done.

soheilhy commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,123 @@
+
+    So from this point of view `e8a61d63` is also not correct.
+
+This is not a theoretical concern: I've found multiple hacks in user
```
nit: There are multiple hacks ...

jacobsa replied:
Done.

soheilhy commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,123 @@
+
+## Proposal
+
+My plan is split into two commits:
```
nit: I would reword this as: The plan is to land this change into steps

jacobsa replied:
Done.

soheilhy commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,123 @@
+    point of this proposal is to make such hacks unnecessary.
+
+2.  Change the interface's implementation of the `Destroy` method to do `delete
+    this`, a more suitable default behavior that frees future subclasses from
```
This can break some random subclass that was relying on the dtor being a no-op.  I believe it needs a major gRPC version bump when we merge the commit.

We should clarify that here in the proposal.

jacobsa replied:
That is in theory also true of the change above; see the last paragraph. Does it need a major bump? Are major version bumps actually done, or does that make this a no-go?

soheilhy replied:
We call "major version" bump as really going from gRPC 1.x to 1.x+1 not gRPC 2. Yes this is something that's done often and there is no issue doing so, once the gRFC is approved.

jacobsa replied:
Okay, I added text saying each should result in a version bump. PTAL.

soheilhy commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,123 @@
+methods. It works just fine once this problem is fixed, but this problem makes
+it extremely awkward otherwise.
+
+The obvious alternative is to leave things as-is. But this requires awful hacks
```
nit: as-is, but this requires ...

jacobsa replied:
Done.

soheilhy reviewed (commented):
Thank you for creating the gRFC.

jacobsa commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,123 @@
+Currently although
+[`grpc::ClientAsyncResponseReaderInterface`](https://github.com/grpc/grpc/blob/7e14d23ab46e1b0924d6c3e40797ed1d587aee7a/include/grpcpp/impl/codegen/async_unary_call.h#L37-L72)
+has a virtual destructor, `std::default_delete` is specialized such that the
+destructor never actually runs. So the only subclasses that can actually be
```
Done.

jacobsa commented:
@soheilhy Thanks for looking. I've addressed all your comments except the one I left open. PTAL.

(Sorry if I'm doing this wrong; I have no idea what conventions are for PRs, so am just treating this like Critique.)

jacobsa commented:
Oh, I also set you to the approver in the doc. Also not sure what the process is there.

soheilhy commented on L79-cpp-async-response-reader-destruction.md:
```diff
@@ -0,0 +1,121 @@
+* Status: Draft
+* Implemented in: C++
+* Last updated: 2021-05-14
+* Discussion at: <google group thread> (filled after thread exists)
```
Before merging the PR, we usually send an email to the mailing list and then merge this PR.

jacobsa replied:
Thanks. Do you want me to start the thread? Can you tell me how/give me an example?

jacobsa replied:
Started the thread and left a link.

soheilhy reviewed (approved):
Please add the group link before merging the PR.

soheilhy reviewed (changes requested):
Sorry I am confused. I shouldn't have clicked approve.

We need two changes:
1) The title of the PR should change to include the suffix L79: ...
2) We need to post this PR to the group and post the thread link here for discussion.

jacobsa commented:
Thanks, both done.

soheilhy commented:
> Thanks, both done.

All look great thanks!

jacobsa commented:
@soheilhy: I believe the 10 business days timeout has been hit. Are you okay with merging this now?

soheilhy reviewed (changes requested):
@jacobsa Can you please update it to L80? Looks like L79 is already merged. Thanks!

jacobsa commented:
Done; it's now L80. I don't know how to "resolve" this comment in github though. :-/

jacobsa commented:
@soheilhy Sorry, meant to tag you before. As far as I can tell you need to clear the "requested changes" bit.

soheilhy reviewed (approved):
@jacobsa thanks for the update! yes, I have to click approve to make it ready to submit.
